bkrdoc/analysis/linter/l_pair_functions.py

```diff
@@ -0,0 +1,121 @@
+    :param before: List of commands (if any) which should be preceded by a Closing command if the Opening one was present,
+                   i.e. Start -> Stop -> Restore;  Restore should not appear before Stop
+    :param each_needs_match: Bool value stating whether each Opening command needs a Closing one or
+                             a single Closing command closes all Opening commands (with matching flag)
+    :param flag_source: String defining an attribute value of Namespace where the flag/name is located, can be None
+    :param flag: Concrete flag of a command whose pair with matching flag is searched for
+    :param restores_all: a flagless Closing command closes everything regardless of flag
+    """
+    def __init__(self, func, pair, before=[], each=False, flag_source=None, restores_all=False):
+        self.func = func
+        self.pair = pair
+        self.before = before
+        self.each_needs_match = each
+        self.restores_all = restores_all
+        self.flag_source = flag_source
+        self.flag = None
+
+    def set_flag(self, flag):
+        self.flag = flag
+
+
+class LinterPairFunctions:
+    """Does the logical analysis of matching paired commands."""
+
+    start_phase_names = ['rlPhaseStart', 'rlPhaseStartTest', 'rlPhaseStartSetup', 'rlPhaseStartCleanup']
+
+    pairs = {'rlPhaseStart': Match('rlPhaseStart', 'rlPhaseEnd', before=start_phase_names, each=True),
+             'rlPhaseStartTest': Match('rlPhaseStartTest', 'rlPhaseEnd', before=start_phase_names, each=True),
+             'rlPhaseStartSetup': Match('rlPhaseStartSetup', 'rlPhaseEnd', before=start_phase_names, each=True),
+             'rlPhaseStartCleanup': Match('rlPhaseStartCleanup', 'rlPhaseEnd', before=start_phase_names, each=True),
+             'rlFileBackup': Match('rlFileBackup', 'rlFileRestore', flag_source='namespace'),
+             'rlVirtualXStart': Match('rlVirtualXStart', 'rlVirtualXStop', flag_source='name'),
+             'rlServiceStart': Match('rlServiceStart', 'rlServiceStop', before=['rlServiceRestore'], flag_source='service')}
+    # Pair(func = 'rlSEBooleanOn', pair = 'rlSEBooleanRestore'), # boolean(s)
+    # Pair(func = 'rlSEBooleanOff', pair = 'rlSEBooleanRestore') } # restore potentially all w/ no arguments given
+
+    currently_unmatched = []
+    errors = []
+
+    def __init__(self):
+        self.currently_unmatched = []
+        self.errors = []
+
+    def analyse(self, list):
+
+        for line in list:
+            match = self.get_match(line.argname)
+            line_flags = self.get_flag(line, match)
+            if type(line_flags) == type([]):
+                for x in line_flags:
+                    setattr(line, match.flag_source, x)
+                    self.analyse_single_line(line)
+            else:
+                self.analyse_single_line(line)
+
+        for elem in self.currently_unmatched:
+            self.add_error(elem.func + " without a matching " + elem.pair, flag=elem.flag)
```

Comment:
For better string formation is better to use .format. "{0} without a matching {1}".format(elem.func, elem.pair). It's just recommendation